File: crates/bitwarden-uniffi/binding-integration-tests/kotlin/SimpleLoginLive.kt
import kotlinx.coroutines.runBlocking
import uniffi.bitwarden_alias.AliasClientSettings
import uniffi.bitwarden_alias.AliasId
import uniffi.bitwarden_alias.ContactId
import uniffi.bitwarden_alias.CreateCustomAliasRequest
import uniffi.bitwarden_alias.CreateRandomAliasRequest
import uniffi.bitwarden_alias.SearchAliasesRequest
import uniffi.bitwarden_uniffi.AliasClient
import uniffi.bitwarden_uniffi.AliasUpdateRequest
import uniffi.bitwarden_uniffi.OptionalSensitiveStringUpdate
import java.util.UUID

fun main() = runBlocking {
    val baseUrl = requireNotNull(System.getenv("SIMPLELOGIN_API_URL"))
    val apiToken = requireNotNull(System.getenv("SIMPLELOGIN_API_TOKEN"))
    val unique = UUID.randomUUID().toString().lowercase()
    val hostname = "kotlin-$unique.integration.test"
    val searchText = "kotlin-live-$unique"
    val client = AliasClient(AliasClientSettings(baseUrl = baseUrl, apiToken = apiToken))
    val aliasIds = mutableListOf<AliasId>()
    var contactId: ContactId? = null

    try {
        val options = client.getAliasOptions(hostname)
        check(options.canCreate && options.suffixes.isNotEmpty())

        val mailboxes = client.listMailboxes()
        val mailbox = mailboxes.firstOrNull { it.verified && it.isDefault }
            ?: mailboxes.firstOrNull { it.verified }
            ?: error("the disposable account must have a verified mailbox")

        val alias = client.createRandomAlias(
            CreateRandomAliasRequest(hostname = hostname, mode = null, note = searchText),
        )
        aliasIds += alias.id
        check(alias.id > 0uL)

        check(client.listAliases(0u, null).aliases.any { it.id == alias.id })
        check(
            client.searchAliases(SearchAliasesRequest(searchText, 0u, null)).aliases
                .any { it.id == alias.id },
        )
        check(client.getAlias(alias.id).email == alias.email)

        val updated = client.updateAlias(
            alias.id,
            AliasUpdateRequest(
                note = OptionalSensitiveStringUpdate.Set("updated-$searchText"),
                name = OptionalSensitiveStringUpdate.Set("Kotlin live alias"),
                mailboxIds = null,
                disablePgp = null,
                pinned = true,
            ),
        )
        check(updated.id == alias.id && updated.name == "Kotlin live alias")
        val cleared = client.updateAlias(
            alias.id,
            AliasUpdateRequest(
                note = OptionalSensitiveStringUpdate.Clear,
                name = null,
                mailboxIds = null,
                disablePgp = null,
                pinned = null,
            ),
        )
        check(cleared.note == null)

        check(!client.disableAlias(alias.id).enabled)
        check(client.enableAlias(alias.id).enabled)
        check(client.setAliasEnabled(alias.id, true).enabled)
        check(client.getAliasRecommendation(hostname)?.alias == alias.email)

        val reverse = client.createReverseAlias(alias.id, "kotlin-contact-$unique@example.net")
        contactId = reverse.id
        check(reverse.id > 0uL)
        check(client.listContacts(alias.id, 0u).contacts.any { it.id == reverse.id })
        check(client.listReverseAliases(alias.id, 0u).aliasId == alias.id)
        check(client.toggleContactBlocked(reverse.id).blockForward)
        check(client.deleteContact(reverse.id).deleted)
        contactId = null

        check(client.listDomains().isNotEmpty())
        client.listCustomDomains()

        val suffix = options.suffixes.firstOrNull { !it.isPremium }
            ?: error("the disposable account must expose a non-premium suffix")
        val custom = client.createCustomAlias(
            CreateCustomAliasRequest(
                aliasPrefix = "kotlin${System.currentTimeMillis()}",
                signedSuffix = suffix.signedSuffix,
                mailboxIds = listOf(mailbox.id),
                hostname = null,
                note = "Kotlin custom alias",
                name = "Kotlin custom",
            ),
        )
        aliasIds += custom.id
        check(custom.id != alias.id)
    } finally {
        contactId?.let {
            try {
                client.deleteContact(it)
            } catch (_: Exception) {
                // Best-effort cleanup after a preceding assertion or transport failure.
            }
        }
        aliasIds.asReversed().forEach {
            try {
                client.deleteAlias(it)
            } catch (_: Exception) {
                // Best-effort cleanup after a preceding assertion or transport failure.
            }
        }
        client.close()
    }

    println("Kotlin SimpleLogin alias lifecycle passed")
}

File: crates/bitwarden-uniffi/binding-integration-tests/swift/SimpleLoginLive.swift
import Foundation

@main
struct SimpleLoginLive {
    static func main() async throws {
        let environment = ProcessInfo.processInfo.environment
        guard
            let baseUrl = environment["SIMPLELOGIN_API_URL"],
            let apiToken = environment["SIMPLELOGIN_API_TOKEN"]
        else {
            fatalError("SIMPLELOGIN_API_URL and SIMPLELOGIN_API_TOKEN are required")
        }

        let unique = UUID().uuidString.lowercased()
        let hostname = "swift-\(unique).integration.test"
        let searchText = "swift-live-\(unique)"
        let client = try AliasClient(
            settings: AliasClientSettings(baseUrl: baseUrl, apiToken: apiToken)
        )
        var aliasIds: [AliasId] = []
        var contactId: ContactId?

        do {
            let options = try await client.getAliasOptions(hostname: hostname)
            precondition(options.canCreate && !options.suffixes.isEmpty)

            let mailboxes = try await client.listMailboxes()
            guard let mailbox = mailboxes.first(where: { $0.verified && $0.isDefault })
                ?? mailboxes.first(where: { $0.verified })
            else {
                fatalError("the disposable account must have a verified mailbox")
            }

            let alias = try await client.createRandomAlias(
                request: CreateRandomAliasRequest(
                    hostname: hostname,
                    mode: nil,
                    note: searchText
                )
            )
            aliasIds.append(alias.id)
            precondition(alias.id > 0)

            let listed = try await client.listAliases(page: 0, filter: nil)
            precondition(listed.aliases.contains(where: { $0.id == alias.id }))
            let searched = try await client.searchAliases(
                request: SearchAliasesRequest(query: searchText, page: 0, filter: nil)
            )
            precondition(searched.aliases.contains(where: { $0.id == alias.id }))
            let detail = try await client.getAlias(aliasId: alias.id)
            precondition(detail.email == alias.email)

            let updated = try await client.updateAlias(
                aliasId: alias.id,
                request: AliasUpdateRequest(
                    note: .set(value: "updated-\(searchText)"),
                    name: .set(value: "Swift live alias"),
                    mailboxIds: nil,
                    disablePgp: nil,
                    pinned: true
                )
            )
            precondition(updated.id == alias.id && updated.name == "Swift live alias")
            let cleared = try await client.updateAlias(
                aliasId: alias.id,
                request: AliasUpdateRequest(
                    note: .clear,
                    name: nil,
                    mailboxIds: nil,
                    disablePgp: nil,
                    pinned: nil
                )
            )
            precondition(cleared.note == nil)

            let disabled = try await client.disableAlias(aliasId: alias.id)
            precondition(!disabled.enabled)
            let enabled = try await client.enableAlias(aliasId: alias.id)
            precondition(enabled.enabled)
            let unchanged = try await client.setAliasEnabled(aliasId: alias.id, enabled: true)
            precondition(unchanged.enabled)
            let recommendation = try await client.getAliasRecommendation(hostname: hostname)
            precondition(recommendation?.alias == alias.email)

            let reverse = try await client.createContact(
                aliasId: alias.id,
                contact: "swift-contact-\(unique)@example.net"
            )
            contactId = reverse.id
            precondition(reverse.id > 0)
            let contacts = try await client.listContacts(aliasId: alias.id, page: 0)
            precondition(contacts.contacts.contains(where: { $0.id == reverse.id }))
            let reverseAliases = try await client.listReverseAliases(aliasId: alias.id, page: 0)
            precondition(reverseAliases.aliasId == alias.id)
            let blocked = try await client.toggleContactBlocked(contactId: reverse.id)
            precondition(blocked.blockForward)
            let contactDeleted = try await client.deleteContact(contactId: reverse.id)
            precondition(contactDeleted.deleted)
            contactId = nil

            let domains = try await client.listDomains()
            precondition(!domains.isEmpty)
            _ = try await client.listCustomDomains()

            guard let suffix = options.suffixes.first(where: { !$0.isPremium }) else {
                fatalError("the disposable account must expose a non-premium suffix")
            }
            let custom = try await client.createCustomAlias(
                request: CreateCustomAliasRequest(
                    aliasPrefix: "swift\(Int(Date().timeIntervalSince1970 * 1_000))",
                    signedSuffix: suffix.signedSuffix,
                    mailboxIds: [mailbox.id],
                    hostname: nil,
                    note: "Swift custom alias",
                    name: "Swift custom"
                )
            )
            aliasIds.append(custom.id)
            precondition(custom.id != alias.id)
        } catch {
            await cleanup(client: client, aliasIds: aliasIds, contactId: contactId)
            throw error
        }

        await cleanup(client: client, aliasIds: aliasIds, contactId: contactId)
        print("Swift SimpleLogin alias lifecycle passed")
    }

    private static func cleanup(
        client: AliasClient,
        aliasIds: [AliasId],
        contactId: ContactId?
    ) async {
        if let contactId {
            _ = try? await client.deleteContact(contactId: contactId)
        }
        for aliasId in aliasIds.reversed() {
            _ = try? await client.deleteAlias(aliasId: aliasId)
        }
    }
}
